### PR TITLE
[SYCLomatic] Refine the migration of cudaExtent/cudaPos

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/APINamesTemplateType.inc [hunk 1]
```diff
@@ -38,6 +38,13 @@ TYPE_REWRITE_ENTRY(
     TYPE_FACTORY(STR(MapNames::getDpctNamespace() + "constant_iterator"),
                      TEMPLATE_ARG(0))))
 
+
+TYPE_REWRITE_ENTRY(
+    "cudaExtent",
+    HEADER_INSERTION_FACTORY(HeaderType::HT_DPL_Iterator,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINamesTemplateType.inc]: while it is DPL related?  Please double check it.

Comment (yihwang-nv) [reply clang/lib/DPCT/APINamesTemplateType.inc]: done

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -15165,3 +15165,139 @@ void TemplateSpecializationTypeLocRule::runRule(
...
+      CSR = CharSourceRange(R, false);
+    }
+    std::string Replacement = "(0, 0, 0)";
+    DpctGlobalInfo::getInstance().addReplacement(
+        std::make_shared<ExtReplacement>(
+            SM, CSR, Replacement, new InsertText(CSR.getBegin(), "(0, 0, 0)")));
+    return;
+  }
+
+  // struct Foo { cudaExtent a; }; -> struct Foo { syc;:range<3> a{0, 0, 0}; };
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]: Fix typo

Comment (yihwang-nv) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 3]
```diff
@@ -69,6 +81,31 @@ std::shared_ptr<TypeLocRewriterFactoryBase> createTypeLocRewriterFactory(
...
+    std::function<TemplateArgsT(const TypeLoc)>... TAsCreator) {
+  return std::make_shared<
+      TypeLocRewriterFactory<CtadTemplateTypeLocRewriter<TypeNameT, TemplateArgsT...>,
+                             std::function<TypeNameT(const TypeLoc)>,
+                             std::function<TemplateArgsT(const TypeLoc)>...>>(
+      std::forward<std::function<TypeNameT(const TypeLoc)>>(TypeNameCreator),
+      std::forward<std::function<TemplateArgsT(const TypeLoc)>>(TAsCreator)...);
+}
+
+// Print a type with no template.
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: without

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 4]
```diff
@@ -69,6 +81,31 @@ std::shared_ptr<TypeLocRewriterFactoryBase> createTypeLocRewriterFactory(
       std::forward<std::function<TypeNameT(const TypeLoc)>>(TypeNameCreator));
 }
 
+// Print a templated type. Pass a STR("") as a template argument for types with
+// no template argument e.g. MyType<>, if enable ctad, use a C++17 class
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: empty template argument e.g. MyType<>. If --enable-ctad is set, the template arguments which could be deduced with class template argument deduction(CTAD) will be omitted in the generated code.

Comment (yihwang-nv) [reply clang/lib/DPCT/TypeLocRewriters.cpp]: done

File: clang/lib/DPCT/TypeLocRewriters.cpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/TypeLocRewriters.cpp]: done

File: clang/test/dpct/cudaExtent.cu [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  // CHECK: sycl::range<3> f = sycl::range<3>({1, 2, 3});
+  cudaExtent f = cudaExtent({1, 2, 3});
+
+  // CHECK: sycl::range<3> g = sycl::range<3>({1, 2, 0});
+  cudaExtent g = cudaExtent({1, 2});
+
+  // CHECK: Bar(sycl::range<3> a) {}
+  Bar(cudaExtent a) {}
+
+  Bar(size_t i, size_t j, size_t k) : a({i, j, k}) {}
```
Comment (ziranzha) [inline clang/test/dpct/cudaExtent.cu]: Consider case like `a{i, j, k}`.

Comment (yihwang-nv) [reply clang/test/dpct/cudaExtent.cu]: done

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**